### PR TITLE
Add Singleton States

### LINKED ISSUE #4793: Add support for "singleton" states
Most states in our merkle tree are map-based, using `MerkleMap` or `VirtualMap`. But many states are "singleton" leafs -- just a single leaf. For example, the `MerkleNetworkContext` or the `RecordsRunningHashLeaf`.  These also must be supported in our state system.

Each of these "singleton" states need to be associated with a state key, be accessible from a `ReadableStates` or `WritableStates`, have lifecycle managed through `Schema`s, and "buffer up" changes until they are finally committed, just like their *KVState peers.

- [x] Create new SPI for `ReadableSingletonState` and `WritableSingletonState`
- [x] Add `getSingleton` to `ReadableStates` and `WritableStates`
- [x] Write tests and update test fixtures to support these
- [x] Update `MerkleHederaState` to support these changes to `ReadableStates` and `WritableStates`
- [x] Add `SingletonNode` which is a `Labeled` merkle node
- [x] Update `MerkleHederaState` to support `SingletonNode`s
- [x] Update schema migration code to support `singleton` nodes
- [x] Add serialization tests for the `SingletonNode`
- [x] Add unit tests for `SingletonNode` and associated classes
- [x] Add Javadoc

### PR DESCRIPTION
**Pull request by rbair23**
Adds singleton states. The PR is a little on the large size, but represents only this one change, of adding singleton states. The following items were covered:

- [x] Create new SPI for `ReadableSingletonState` and `WritableSingletonState`
- [x] Add `getSingleton` to `ReadableStates` and `WritableStates`
- [x] Write tests and update test fixtures to support these
- [x] Update `MerkleHederaState` to support these changes to `ReadableStates` and `WritableStates`
- [x] Add `SingletonNode` which is a `Labeled` merkle node
- [x] Update `MerkleHederaState` to support `SingletonNode`s
- [x] Update schema migration code to support `singleton` nodes
- [x] Add serialization tests for the `SingletonNode`
- [x] Add unit tests for `SingletonNode` and associated classes
- [x] Add Javadoc

Closes #4793

### CODE REVIEW

**Review by tinker-michaelj — DISMISSED** [review]
LGTM!

File: hedera-node/hedera-app-spi/src/main/java/com/hedera/node/app/spi/state/ReadableSingletonState.java [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+     */
+    @Nullable
+    T get();
+
+    /**
+     * Gets whether the value of this {@link ReadableSingletonState} has been read.
+     *
+     * @return true if {@link #get()} has been called on this instance
+     */
+    boolean read();
```
Comment (Neeharika-Sompalli) [inline hedera-node/hedera-app-spi/src/main/java/com/hedera/node/app/spi/state/ReadableSingletonState.java:55]: Probably should be `isRead` ?

File: hedera-node/hedera-app-spi/src/main/java/com/hedera/node/app/spi/state/ReadableSingletonStateBase.java [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+
+/**
+ * A convenient implementation of {@link ReadableSingletonStateBase}.
+ *
+ * @param <T> The type of the value
+ */
+public class ReadableSingletonStateBase<T> implements ReadableSingletonState<T> {
+    private final String stateKey;
+    private final Supplier<T> backingStoreAccessor;
+    private boolean read = false;
```
Comment (Neeharika-Sompalli) [inline hedera-node/hedera-app-spi/src/main/java/com/hedera/node/app/spi/state/ReadableSingletonStateBase.java:30]: nit : similarly to the variable

File: hedera-node/hedera-app-spi/src/main/java/com/hedera/node/app/spi/state/StateDefinition.java [hunk 3]
```diff
@@ -33,7 +34,77 @@
...
+     *
+     * @param stateKey The state key
+     * @param valueSerdes The serdes for the singleton value
+     * @return An instance of {@link StateDefinition}
+     * @param <K> The key type
+     * @param <V> The value type
+     */
+    public static <K extends Comparable<K>, V> StateDefinition<K, V> singleton(
+            @NonNull final String stateKey, @NonNull final Serdes<V> valueSerdes) {
+        return new StateDefinition<>(stateKey, null, valueSerdes, 1, false, true);
```
Comment (Neeharika-Sompalli) [inline hedera-node/hedera-app-spi/src/main/java/com/hedera/node/app/spi/state/StateDefinition.java:108]: Sorry if this is obvious. But why is `maxKeysHint` 1 here ? Shouldn't it be set only for `onDisk` singletons ?

File: hedera-node/hedera-app-spi/src/main/java/com/hedera/node/app/spi/state/WritableSingletonStateBase.java [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+    }
+
+    @Override
+    public void put(T value) {
+        this.modified = true;
+        this.value = value;
+    }
+
+    @Override
+    public boolean modified() {
```
Comment (Neeharika-Sompalli) [inline hedera-node/hedera-app-spi/src/main/java/com/hedera/node/app/spi/state/WritableSingletonStateBase.java:69]: Should be `isModified()` ?

File: hedera-node/hedera-app/src/main/java/com/hedera/node/app/state/merkle/singleton/StringLeaf.java [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+package com.hedera.node.app.state.merkle.singleton;
+
+import com.swirlds.common.io.streams.SerializableDataInputStream;
+import com.swirlds.common.io.streams.SerializableDataOutputStream;
+import com.swirlds.common.merkle.MerkleLeaf;
+import com.swirlds.common.merkle.impl.PartialMerkleLeaf;
+import com.swirlds.common.utility.Labeled;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import java.io.IOException;
+
```
Comment (Neeharika-Sompalli) [inline hedera-node/hedera-app/src/main/java/com/hedera/node/app/state/merkle/singleton/StringLeaf.java:25]: Javadocs for the class on its usage will be helpful.

File: hedera-node/hedera-app/src/main/java/com/hedera/node/app/state/merkle/singleton/ValueLeaf.java [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+import com.swirlds.common.io.streams.SerializableDataInputStream;
+import com.swirlds.common.io.streams.SerializableDataOutputStream;
+import com.swirlds.common.merkle.MerkleLeaf;
+import com.swirlds.common.merkle.impl.PartialMerkleLeaf;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import edu.umd.cs.findbugs.annotations.Nullable;
+import java.io.DataInputStream;
+import java.io.IOException;
+import java.util.Objects;
+
```
Comment (Neeharika-Sompalli) [inline hedera-node/hedera-app/src/main/java/com/hedera/node/app/state/merkle/singleton/ValueLeaf.java:28]: same here. Javadoc for the class on its usage will be helpful.

**Review by Neeharika-Sompalli — DISMISSED** [review]
LGTM other than few questions and nits.

File: hedera-node/hedera-app-spi/src/main/java/com/hedera/node/app/spi/state/StateDefinition.java (hunk 3 above)
Comment (rbair23) [reply hedera-node/hedera-app-spi/src/main/java/com/hedera/node/app/spi/state/StateDefinition.java:108]: No reason at all, TBH. The field is ignored, but if it *were* to have a meaning, I thought "1" was true. But I don't think it matters. Maybe -1 would be better so nobody thinks there is a meaning to it.

File: hedera-node/hedera-app/src/main/java/com/hedera/node/app/state/merkle/singleton/StringLeaf.java (hunk 5 above)
Comment (rbair23) [reply hedera-node/hedera-app/src/main/java/com/hedera/node/app/state/merkle/singleton/StringLeaf.java:25]: Done

File: hedera-node/hedera-app/src/main/java/com/hedera/node/app/state/merkle/singleton/ValueLeaf.java (hunk 6 above)
Comment (rbair23) [reply hedera-node/hedera-app/src/main/java/com/hedera/node/app/state/merkle/singleton/ValueLeaf.java:28]: Added some docs

File: hedera-node/hedera-app-spi/src/main/java/com/hedera/node/app/spi/state/StateDefinition.java (hunk 3 above)
Comment (rbair23) [reply hedera-node/hedera-app-spi/src/main/java/com/hedera/node/app/spi/state/StateDefinition.java]: Changed to -1 (with a static constant `NO_MAX`)

File: hedera-node/hedera-app-spi/src/main/java/com/hedera/node/app/spi/state/ReadableSingletonState.java (hunk 1 above)
Comment (rbair23) [reply hedera-node/hedera-app-spi/src/main/java/com/hedera/node/app/spi/state/ReadableSingletonState.java]: Done

**Review by Neeharika-Sompalli — APPROVED** [review]
LGTM ! 💯

**Pull request merged**